### ISSUE TITLE
Does .allowframebreaks conflict with \hypertarget?

### ISSUE THREAD

**Issue by lifepillar** [jgm/pandoc#2704]
I have a Markdown document containing several citations, which occupy more or less two Beamer slides. When I typeset the document to produce a PDF, I get two slides at the end, called “References I” and “References II”, but the former is empty and the latter flows out. The problem disappears if I manually remove the `\hypertarget` commands that Pandoc puts before each reference. Is there a way to suppress those commands? I have set `link-citations: false`, but to no avail.

Attached you find a minimal example showing the problem.

[biblio.zip](https://github.com/jgm/pandoc/files/121827/biblio.zip)

**Comment by lalthomas** [comment]
I have the same issue

**Comment by etc** [comment]
Also having this issue.

**Comment by etc** [comment]
@lifepillar and @lalthomas —did either of you ever find a fix for this?

**Comment by lifepillar** [comment]
@etc Unfortunately, no. I am now using ConTeXt to great satisfaction. Really, the simplicity of Markdown is appealing at first, but it shows its limits as soon as you need a little more control over any aspect of your documents. If you know (some variant of) TeX, it's likely that nothing can be more productive than writing documents in TeX.

To be clear, I am not complaining about pandoc, which is incredibly great!

**Comment by wilx** [comment]
This appears to caused by widow penalties in the generated LaTeX file:

```latex
% Prevent slide breaks in the middle of a paragraph:
\widowpenalties 1 10000
```

If you comment out this line, it will split the frames.

**Comment by wilx** [comment]
FYI, [this answer](http://tex.stackexchange.com/a/21985/28495) to [How to avoid page-breaks inside paragraphs? - TeX - LaTeX Stack Exchange](http://tex.stackexchange.com/questions/21983/how-to-avoid-page-breaks-inside-paragraphs) is related.

I have experimented with the given test case and it seems to me that adding `\widowpenalties 2 10000 0` after the `\begin{frame}` if the references frame creates the most reasonable output.

**Comment by wilx** [comment]
Additionally, [this answer](http://tex.stackexchange.com/a/260729/28495) to [No page break when using hypertarget in custom environment - TeX - LaTeX Stack Exchange](http://tex.stackexchange.com/questions/260704/no-page-break-when-using-hypertarget-in-custom-environment) seems also related and it is supported by the initial observation of removing `\hypertarget` to fix things. From the linked answer:

> This is the “destination” created with the `\hypertarget` command. Now, the `hyperref` package inserts a `\penalty 10000` immediately below the destination, to explicitly prohibit a page break at that point, because such a break would detach the destination anchor from the material it is supposed to link to. Without this, TeX could break the page at the next `\glue` item (the one marked `\parskip`), and indeed you have found that, without the `\hypertarget` command, everything works fine; but, as it stands, the vertical list shown above simply has no permissible breakpoint between paragraphs. If the paragraphs were several lines long, however, there would be legal breakpoints within the paragraphs themselves.

**Comment by mb21** [comment]
If anyone has a specific recommendation how the latex generated by pandoc should be changed, please post it. Preferably make sure it doesn't break any other edge cases...

**Comment by wilx** [comment]
@mb21 According to [David Carlisle](http://chat.stackexchange.com/transcript/message/35781858#35781858), we should insert `\leavevmode` at the start of the paragraph before the `\hypertarget`. I have [tested this on Overleaf](https://www.overleaf.com/read/wbdzgfgmsjtb) and it seems to fix the issue.

**Comment by mb21** [comment]
Inserting `\leavemode` only in the References section and only when generating beamer or..?

**Comment by wilx** [comment]
@mb21 Well, I am not sure. It seems to me that only if `\hypertarget` is the first thing in a paragraph then `\leavevmode` should be issued before it.

**Comment by mb21** [comment]
Would be good if a few more people weigh in on the change... how sure are we this wouldn't break anything?

**Comment by davidcarlisle** [comment]
If you have

```
<blank line>
\special{whatever}%
Paragraph text
```
then the special comes above the paragraph and causes all sorts of spacing and page breaking issues.

If you have 

```
<blank line>
\leavemode\special{whatever}%
Paragraph text
```

then the whatsit comes in the first line of the paragraph and is much safer.

However if the following thing is a vertical construct so instead of `Paragraph text` you have `\begin{center}` or `\[` or `\section...`  then

```
<blank line>
\leavemode\special{whatever}%
\section{zzzz]
```
then you will get a spurious "white" paragraph that looks like an extra vertical space.


You get exactly the same issue with \color and in the end I just had to document that in some places using \color affects the vertical spacing and you can avoid that by using \textcolor (which uses \leavevmode) or by adjusting the location of where you insert \color.


So it is hard (probably impossibly hard) to define \hypertarget to tell from within tex whether to add the whatsit at that point, however (depending on your architecture here?)  perhaps in this specific case you know that you are writing the text of the citation immediately after the `\hypertarget` so you can insert `\leavevmode` at the start.

**Comment by jgm** [comment]
The fix I've committed adds the `\leavevmode` in one
particular case only:  when you have a Div with an id
(which will trigger a hypertarget) and the first block
in the Div is a paragraph.  That at least handles the
pandoc-citeproc case well, without causing any problems
I've yet noticed...

**Comment by etc** [comment]
Fantastic, thanks @jgm

**Comment by JohnLukeBentley** [comment]
> without causing any problems I've yet noticed

`\leavevmode\vadjust pre{}` is causing the entire CSL line (including `\CSLLeftMargin` and `\CSLRightInline`) to shift rightwards, into the right margin, for (apparently) CSL styles with a citation-format of "numeric" or "label". Or, in terms of pandoc latex output, the problem arises when a  `\CSLLeftMargin` is present with `\leavevmode\vadjust pre{}`.

For example, using a "numeric" formatted CSL, [IEEE (with URL)](https://www.zotero.org/styles?q=id%3Aieee-with-url), we get the following:
![image](https://user-images.githubusercontent.com/18357541/118885138-b7edf880-b93a-11eb-86a7-2bca1cdd99d2.png)

If in the intermediate latex that pandoc spits out we remove `\leavevmode\vadjust pre{}` to leave `\hypertarget{lorem}{}` , and compile the latex directly, then beauty is restored:

Handy regex:

* find: `^\\leavevmode\\vadjust pre\{(.*)\}`
* replace: `\1`

![image](https://user-images.githubusercontent.com/18357541/118886070-d56f9200-b93b-11eb-84ef-44a2e7a2830c.png)

Similar behaviour can be observed with [DIN 1505-2 (alphanumeric, Deutsch) - standard superseded by ISO-690](https://www.zotero.org/styles?q=din-1505-2-alphanumeric), (apparently) the only "label" format type CSL at Zotero Style Repository.

The problem emerges in my case when using a custom CSL, based on [Chicago Manual of Style 17th edition (author-date)](https://www.zotero.org/styles?q=id%3Achicago-author-date) - a style with citation-format="author-date" - which exploits shorttitle as a label. And, at the pandoc end, I redefine `\CSLLeftMargin` and `\CSLRightInline` to do horizontal sizing adjustments, to fit labels (`\CSLLeftMargin`) of varying widths. Where the problem becomes even more pronounced visually:

![image](https://user-images.githubusercontent.com/18357541/118887540-dacddc00-b93d-11eb-9d58-5346ec00d004.png)

Observe that the regular author-date references above, those without a label and starting at "Chalmers", a formatted OK. In the latex we have ...

```
\leavevmode\vadjust pre{\hypertarget{ref-chalmers_1995_facinga}{}}%
\CSLRightInline{Chalmers,
```
... that is, no `\CSLLeftMargin`.

I'm currently unable to suggest a fix.

Whatever the fix, I'd request that a simple conditional to remove `\leavevmode\vadjust pre{` for some types of CSL citation-format(s) be not done. In particular, I wouldn't want CSL styles with citation-format="author-date" to retain the problem.

**Comment by jgm** [comment]
I do see the large space in ieee; however, I don't notice any change when I remove `\leavevmode\vadjust pre`.
I haven't tried with the label style, but in ieee I see no difference. There, the spacing is caused by the CSLLeftMargin macro, which uses a fixed width of `\csllabelwidth`. If you redefine this you can get a narrower spacing. Try putting this in the tex file (after the CSL definitions):
```
\setlength{\csllabelwidth}{1em}
```

**Comment by JohnLukeBentley** [comment]
But in ieee:
* Do you recognize each reference line is shifted rightwards into the right margin in my first screenshot (In my second screenshot that problem doesn't exist)? See in the first reference the string "Journal of Con-" relative to the above lorem paragraph and the top margin horizontal rule (which extends all the way to the side margins); and
* Can you reproduce this?

I'm on pandoc v2.13.

**Comment by JohnLukeBentley** [comment]
As you were. I had thought I was outputting my ieee example *without* dong the CSL redefining. However, when I remove my CSL redefining the problem disappears. `\leavevmode\vadjust pre{` is causing me issues which are readily resolved if removed. However, I'll press on with further debugging of my own code over the next day or so, including taking a look at your helpful suggestion to redefine `\csllabelwidth`.

**Comment by jgm** [comment]
I see the shift now (I had thought you were asking about the wide margin of the label).  But I can't reproduce it.   Default pandoc does not show this using ieee.  It must be an interaction with something you're doing, presumably in a custom template?

I suggest you open up a new issue about this, including the inputs needed to reproduce it.

**Comment by jgm** [comment]
Probably the issue is with your redefinition of the CSL environments...

**Comment by JohnLukeBentley** [comment]
For completeness, this did turn out to be an issue with my code.

I made two errors:

* I failed to end my latex outputting redefinition commands of `\CSLLeftMargin` and `\CSLRightInline`, with a comment character `%`. And so additional spaces where inserted in my output. E.g. this is proper form ...

      \newcommand{\gammapart}[1]{%
        \mbox{#1}%
      }
* And, relatedly, I failed to recognize that pandoc's bibliographic entries where placing a space between `\CSLLeftMargin{ ... }`  `\CSLRightInline{ ...}`, in virtue of not putting a comment character `%` at the end of the `\CSLLeftMargin{ ... }` output. E.g.

      \CSLLeftMargin{\textbar Z\textbar{} }
      \CSLRightInline{Nietzsche, Friedrich ...
